### ISSUE TITLE
Replace DependencyPath with Guava Graph

### ISSUE THREAD

**Issue by suztomo** [GoogleCloudPlatform/cloud-opensource-java#1405]
Is it feasible to replace current tree representation of `List<DependencyPath>` with Guava's Graph?

# Considerations

In Maven dependency graph edges are ordered.

Dependency tree represented by `DependencyPath`s are easy to get the path from the root to the leaf (`DependencyPath` itself). Maybe the node can have edges path to root (or parents). This is the biggest difference from the Maven-resolver's native dependency tree, which consists of `DependnecyNode`s. A DependencyNode does not know its parent.

For BOM Dashboard, some existing logic for dependency convergence (irrelevant to Linkage Checker) columns is based on DependencyPath.  Excerpt from `testJaxen`
```
    List<Update> updates = graph.findUpdates();
    Truth.assertThat(updates).hasSize(6);

    List<DependencyPath> conflicts = graph.findConflicts();
    Truth.assertThat(conflicts).hasSize(10);
```

Should we migrate that logic too? Or limit the scope of this PR to the graph used by Linkage Checker.

The OutOfMemoryError in #1256 drives this change. We have to ensure the resulting design would pass Beam’s hcatalog artifact.

# Design

Maven-resolver's `DependencyNode` has Maven artifact, scope, and optional flag.

**Comment by elharo** [comment]
1. We definitely don't want two graph representations.
 
2. We do need to track and store the paths to the dependencies. That's the biggest reason we built this unusual representation. That's also probably why we have the OOMEs. Possibly we can figure out how to stream the processing so we don't have to hold it all in memory at once.

3. This should be a pure refactoring that does not change any tests. 

4. We shouldn't have to change the public API to change the graph representation. However, currently we expose a lot of internal implementation details in multimaps and lists and the like we pass from one class to another so we might have to improve our data encapsulation first before we can adequately explore this.

5. If we work on this, we should time box it. We don't want to burn all our developer cycles on a speculative fix for one bug. 

6. We did consider these classes a couple of years ago, and decided against them. Perhaps it's time to revisit that decision, but it wasn't simply a case of not being familiar with them.
